openstack/standalone-cinder/pkg/provisioner/provisioner.go

```diff
@@ -179,6 +190,12 @@ func (p *cinderProvisioner) Delete(pv *v1.PersistentVolume) error {
 
 	mapper.AuthTeardown(p, pv)
 
+	err = p.vsb.detachCinderVolume(p.VolumeService, volumeID)
+	if err != nil {
+		glog.Errorf("Failed to detach volume %s: %v", volumeID, err)
+		return err
+	}
+
```

Comment:
These are great!  One thing I was thinking though; what about condensing these into a single recovery function for Detach related failures?  So just have a general "_handleDetachErr" and check where in the process the volume appears to be and do basically the same steps you have here?

I think it might be cleaner?

One reason I think of this is I've been looking at adding the new Attach API's to Gophercloud and *some day* would like to add them here as well; it would make things easier in the future to add new support.

Reply:
Thanks @j-griffith for the review.  I agree with you that a detach handler could be cleaner and it is something I have considered adding in the future.  This would help handle cases where one step in the cleanup fails and causes subsequent attempts at cleanup to fail.  eg. Detach succeeds but disconnect fails.  Next time we retry detach but it fails because it is already detached.

That being said, I would prefer to merge this batch of PRs first and then address this next.